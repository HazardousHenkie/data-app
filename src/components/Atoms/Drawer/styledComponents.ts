import styled from 'styled-components'

export const HandleBar = styled.div`
    position: fixed;
    border-radius: 100px;
    border: 3px solid ${(props): string => props.theme.black};
    margin: 0 auto 8px;
    width: 50%;
    bottom: 0;
    left: 25%;
    top: auto;
    background: ${(props): string => props.theme.black};
`

export const DrawerWrapper = styled.div`
    position: relative;
    bottom: 0;
`
