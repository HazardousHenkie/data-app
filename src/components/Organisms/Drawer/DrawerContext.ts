import { createContext } from 'react'

interface DrawerContextInterface {
    openDrawer: boolean
    setOpenDrawer: React.Dispatch<React.SetStateAction<boolean>>
}

const DrawerContext = createContext<DrawerContextInterface>({
    openDrawer: false,
    setOpenDrawer: (): void => {}
})

export default DrawerContext
