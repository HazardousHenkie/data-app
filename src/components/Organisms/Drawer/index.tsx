import React, { useState, useEffect } from 'react'

import { SwipeableDrawerProps } from '@material-ui/core/SwipeableDrawer'

import {
    SwipeableDrawerStyled,
    HandleBar,
    DrawerWrapper,
    ClickIndicator,
    SwipeIndicator,
    SwipeIndicatorInner,
    SwipeIndicatorInnerDown
} from './styledComponents'

interface DrawerCustomProps {
    externalOpenDrawer: any
}

const Drawer: React.FC<DrawerCustomProps &
    Omit<SwipeableDrawerProps, 'open' | 'onClose' | 'onOpen'>> = ({
    externalOpenDrawer,
    children
}) => {
    const [openDrawer, setOpenDrawer] = useState(false)

    useEffect(() => {
        if (externalOpenDrawer) {
            setOpenDrawer(true)
        }
    }, [externalOpenDrawer])

    const toggleDrawer = (open: boolean) => (
        event: React.KeyboardEvent | React.MouseEvent
    ) => {
        if (
            event &&
            event.type === 'keydown' &&
            ((event as React.KeyboardEvent).key === 'Tab' ||
                (event as React.KeyboardEvent).key === 'Shift')
        ) {
            return
        }

        setOpenDrawer(open)
    }

    return (
        <>
            <ClickIndicator onClick={toggleDrawer(true)}>
                <SwipeIndicatorInner />
            </ClickIndicator>

            <DrawerWrapper>
                <HandleBar>
                    <SwipeIndicator onClick={toggleDrawer(true)}>
                        <SwipeIndicatorInner />
                    </SwipeIndicator>
                </HandleBar>

                <SwipeableDrawerStyled
                    anchor="bottom"
                    open={openDrawer}
                    onClose={toggleDrawer(false)}
                    onOpen={toggleDrawer(true)}
                >
                    <SwipeIndicator down onClick={toggleDrawer(false)}>
                        <SwipeIndicatorInnerDown />
                    </SwipeIndicator>

                    <HandleBar relative />

                    {children}
                </SwipeableDrawerStyled>
            </DrawerWrapper>
        </>
    )
}

export default Drawer
