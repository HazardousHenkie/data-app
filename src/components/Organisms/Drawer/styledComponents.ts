import styled, { keyframes, Keyframes, css } from 'styled-components'

import ArrowDropUpIcon from '@material-ui/icons/ArrowDropUp'
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown'

import SwipeableDrawer from '@material-ui/core/SwipeableDrawer'

interface HandleBarProps {
    relative?: boolean
}

interface SwipeIndicatorProps {
    down?: boolean
}

const arrow = css`
    position: absolute;
    left: 50%;
    margin-left: -12px;
    color: ${({ theme }) => theme.palette.common.black};
`

const indicator = css`
    cursor: pointer;
    opacity: 0.5;
    position: absolute;
    left: 50%;
    width: 30px;
    height: 30px;
    margin-left: -15px;
    border: 2px solid ${({ theme }) => theme.palette.common.black};
    border-radius: 50px;
`

export const swipeAnimation = (reverse: boolean): Keyframes => keyframes`
    0% {
        transform: translate(0, 0);
        opacity: 0;
    }
    40% {
        opacity: 1;
    }
    80% {
        transform: translate(0, ${reverse ? '10px' : '-10px'});
        opacity: 0;
    }
    100% {
        opacity: 0;
    }
`

export const SwipeableDrawerStyled = styled(SwipeableDrawer)`
    .MuiDrawer-paper {
        padding: 20px;
        border-top-right-radius: 25px;
        border-top-left-radius: 25px;
    }
`

export const HandleBar = styled.div<HandleBarProps>`
    border-radius: 100px;
    width: 50%;
    bottom: 0;
    top: auto;

    ${props =>
        props.relative
            ? `
            position: relative;
            left: 0;
            margin: -5px auto 10px;
        `
            : `
            position: fixed;
            left: 25%;
            margin: 0 auto 8px;
        `}

    ${({ theme }) => `
        background: ${theme.palette.common.black};
        border: 2px solid ${theme.palette.common.black};

        ${theme.breakpoints.up('md')} {
            display: none;
        }
    `}
`

export const DrawerWrapper = styled.div`
    position: relative;
    bottom: 0;
`

export const SwipeIndicator = styled.div<SwipeIndicatorProps>`
    ${indicator}
    ${(props): string => (!props.down ? 'bottom: 10px;' : 'top: -35px;')};
`

export const ClickIndicator = styled.div`
    ${indicator}
    bottom: 10px;

    ${({ theme }) => theme.breakpoints.down('sm')} {
        display: none;
    }
`

export const SwipeIndicatorInner = styled(ArrowDropUpIcon)`
    ${arrow}
    top: 5px;
    animation: ${swipeAnimation(false)} 2s infinite;
`

export const SwipeIndicatorInnerDown = styled(ArrowDropDownIcon)`
    ${arrow}
    top: -5px;
    animation: ${swipeAnimation(true)} 2s infinite;
`
