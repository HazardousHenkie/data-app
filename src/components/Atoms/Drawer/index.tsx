import React, { useState } from 'react'

import SwipeableDrawer, {
    SwipeableDrawerProps
} from '@material-ui/core/SwipeableDrawer'

import { HandleBar, DrawerWrapper } from './styledComponents'

const Drawer: React.FC<Omit<
    SwipeableDrawerProps,
    'open' | 'onClose' | 'onOpen'
>> = ({ children }) => {
    const [openDrawer, setOpenDrawer] = useState(false)

    const toggleDrawer = (open: boolean) => (
        event: React.KeyboardEvent | React.MouseEvent
    ): void => {
        if (
            event &&
            event.type === 'keydown' &&
            ((event as React.KeyboardEvent).key === 'Tab' ||
                (event as React.KeyboardEvent).key === 'Shift')
        ) {
            return
        }

        setOpenDrawer(open)
    }

    return (
        <DrawerWrapper>
            <HandleBar />

            <SwipeableDrawer
                anchor="bottom"
                open={openDrawer}
                onClose={toggleDrawer(false)}
                onOpen={toggleDrawer(true)}
            >
                <HandleBar relative />

                {children}
            </SwipeableDrawer>
        </DrawerWrapper>
    )
}

export default Drawer
