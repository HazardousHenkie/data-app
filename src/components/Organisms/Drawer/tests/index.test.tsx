import React from 'react'
import renderer from 'react-test-renderer'

import MockingAppComponent from 'utils/testing/MockingAppComponent'
import Drawer from '../index'

describe('<Drawer />', () => {
    it('should render like snapshot', () => {
        const component = renderer.create(
            <MockingAppComponent>
                <Drawer />
            </MockingAppComponent>
        )

        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
})
