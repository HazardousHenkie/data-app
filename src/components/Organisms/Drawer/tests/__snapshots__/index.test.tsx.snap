// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> should render like snapshot 1`] = `
Array [
  <div
    className="sc-AxhUy laGaXi"
    onClick={[Function]}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root sc-AxgMl eAPdQb"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 14l5-5 5 5z"
      />
    </svg>
  </div>,
  <div
    className="sc-AxiKw boTYJG"
  >
    <div
      className="sc-AxirZ evChcC"
    >
      <div
        className="sc-AxhCb gvnvGs"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root sc-AxgMl eAPdQb"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 14l5-5 5 5z"
          />
        </svg>
      </div>
    </div>
  </div>,
]
`;
